src/generic.jl

```diff
@@ -104,18 +146,34 @@ function _pairwise!(r::AbstractMatrix, metric::PreMetric,
     r
 end
 
+function _pairwise!(r::AbstractMatrix, metric::SemiMetric, a)
+    require_one_based_indexing(r)
+    n = length(a)
+    size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r."))
+    itr = Iterators.product(enumerate(a), enumerate(a))
+    @inbounds for ((i, ai), (j, aj)) in itr
+        r[i, j] = i > j ? 
+            metric(ai, aj) :
+            i == j ?
+                0 :
+                r[j,i]
```

Comment:
I find this a bit hard to read. How about something like this (many variants are possible)?
```suggestion
        r[i, j] = if i > j
            metric(ai, aj)
        elseif
            0        
        else
            r[j, i]
        end
```